### PR TITLE
Implement Annotator for TLS validation

### LINKED ISSUE #43: Implement Annotator for TLS Validation
The Annotator should be able to validate whether or not the incoming or outgoing HTTP request is being conducted over TLS.

### PR DESCRIPTION
**Pull request by KarimElghamry**
* Add TLS AnnotationType
* Implement TlsAnnotator to validate TLS connection
  and sign the produced annotation
* Add AnnotatorUtils class to govern the common operations
  done by all annotators
* Add TlsAnnotator to the AnnotatorFactory
* Implement related unit test

Fix #43

Signed-off-by: Karim Elghamry <karimelghamry@gmail.com>

### CODE REVIEW

File: src/main/java/com/alvarium/contracts/AnnotationType.java [hunk 1]
```diff
@@ -1,5 +1,10 @@
 package com.alvarium.contracts;
 
+import com.google.gson.annotations.SerializedName;
+
 public enum AnnotationType {
```
Comment (KarimElghamry) [inline src/main/java/com/alvarium/contracts/AnnotationType.java:5]: The rest of the enum serializations are implemented in the PR #46. Once it's merge, I will rebase and resolve conflicts.

**Comment by tsconn23** [comment]
@KarimElghamry Resolve conflicts when you can via rebase, and then I'll take a look at this one.

**Comment by KarimElghamry** [comment]
> 
> 
> @KarimElghamry Resolve conflicts when you can via rebase, and then I'll take a look at this one.

Rebased in 90b81f6c5b8b0cbd6a944fc2898949a6bfdcbd38

**Review by tsconn23 — CHANGES_REQUESTED** [review]

File: src/main/java/com/alvarium/annotators/TlsAnnotator.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    } catch (UnknownHostException e) {
+      throw new AnnotatorException("cannot get host name.", e);
+    }
+
+    // TLS check handshake
+    Boolean isStatified = this.verfiyHandshake(ctx.getProperty(AnnotationType.TLS.name(),
+        SSLSocket.class));
+
+    // create an annotation without signature
+    final Annotation annotation = new Annotation(key, hash, host, kind, null, isStatified,
```
[the remark this answers is not in the document]
Comment (KarimElghamry) [reply src/main/java/com/alvarium/annotators/TlsAnnotator.java]: I don't know what happened when I was typing this. must've been the vim key bindings being messed up inside of vscode. fixed in 00386d7feec721a46ad75d0b380d5db0b109a1fb

File: src/main/java/com/alvarium/annotators/AnnotatorUtils.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+import com.alvarium.hash.HashTypeException;
+import com.alvarium.sign.KeyInfo;
+import com.alvarium.sign.SignException;
+import com.alvarium.sign.SignProvider;
+import com.alvarium.sign.SignProviderFactory;
+import com.alvarium.utils.Encoder;
+
+/**
+ * A Util class responsible for carrying out common operations done by the annotators
+ */
```
[the remark this answers is not in the document]
Comment (KarimElghamry) [reply src/main/java/com/alvarium/annotators/AnnotatorUtils.java:21]: makes sense. I think this approach is cleaner. fixed in 00386d7feec721a46ad75d0b380d5db0b109a1fb

File: src/main/java/com/alvarium/annotators/TlsAnnotator.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    // get host name
+    String host;
+    try {
+      host = InetAddress.getLocalHost().getHostName();
+    } catch (UnknownHostException e) {
+      throw new AnnotatorException("cannot get host name.", e);
+    }
+
+    // TLS check handshake
+    Boolean isStatified = this.verfiyHandshake(ctx.getProperty(AnnotationType.TLS.name(),
```
[the remark this answers is not in the document]
Comment (KarimElghamry) [reply src/main/java/com/alvarium/annotators/TlsAnnotator.java]: fixed in 00386d7feec721a46ad75d0b380d5db0b109a1fb

File: src/main/java/com/alvarium/contracts/Annotation.java [hunk 5]
```diff
@@ -38,6 +38,32 @@ public Annotation(String key, HashType hash, String host, AnnotationType kind, S
...
+    }
+
+    /**
+     * a constructor that returns a copy of this object and replacing the signature with 
+     * the provided one
+     * @param sig
+     * @return
+     */
+    public Annotation withSignature(String sig) {
+      return new Annotation(this.id, this.key, this.hash, this.host, this.kind, sig, this.isSatisfied
```
[the remark this answers is not in the document]
Comment (KarimElghamry) [reply src/main/java/com/alvarium/contracts/Annotation.java]: I was aiming for absolute immutability. Yea, I think a normal setter would be enough in our usecase. setter added in 00386d7feec721a46ad75d0b380d5db0b109a1fb

**Review by tsconn23 — APPROVED** [review]

**Pull request merged**